applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py

```diff
@@ -0,0 +1,150 @@
+
+        # Initialize output interval data
+        self.next_output = 0.0
+
+        # Initialize the snapshots data list
+        self.snapshots_data_list = []
+
+    @classmethod
+    def GetDefaultParameters(self):
+        default_settings = KratosMultiphysics.Parameters("""{
+            "help": "A process to set the snapshots matrix and calculate the ROM basis from it.",
+            "model_part_name": "",
+            "snapshots_control_type": "step",
+            "snapshots_interval": 1.0,
+            "nodal_unknowns": [],
+            "rom_basis_output_format": "json",
+            "rom_basis_output_name": "RomParameters",
+            "svd_truncation_tolerance": 1.0e-6
+        }""")
+
+        return default_settings
+
+    def IsOutputStep(self):
+        if self.snapshots_control_is_time:
+            time = self.__GetPrettyFloat(self.model_part.ProcessInfo[KratosMultiphysics.TIME])
+            return time >= self.__GetPrettyFloat(self.next_output)
+        else:
+            step = self.__GetPrettyFloat(self.model_part.ProcessInfo[KratosMultiphysics.STEP])
+            return step >= self.next_output
+
+    def PrintOutput(self):
+        # Save the data in the snapshots data list
+        aux_data_array = []
+        for node in self.model_part.Nodes:
+            for snapshot_var in self.snapshot_variables_list:
+                aux_data_array.append(node.GetSolutionStepValue(snapshot_var))
+        self.snapshots_data_list.append(aux_data_array)
+
+        # Schedule next snapshot output
+        if self.snapshots_interval > 0.0: # Note: if == 0, we'll just always print
+            if self.snapshots_control_is_time:
+                time = self.__GetPrettyFloat(self.model_part.ProcessInfo[KratosMultiphysics.TIME])
+                while self.__GetPrettyFloat(self.next_output) <= time:
+                    self.next_output += self.snapshots_interval
+            else:
+                step = self.__GetPrettyFloat(self.model_part.ProcessInfo[KratosMultiphysics.STEP])
+                while self.next_output <= step:
+                    self.next_output += self.snapshots_interval
+
+    def ExecuteFinalize(self):
+        # Set a NumPy array with the snapshots data
+        n_nodes = self.model_part.NumberOfNodes()
+        n_data_cols = len(self.snapshots_data_list)
+        n_nodal_unknowns = len(self.snapshot_variables_list)
+        snapshots_matrix = numpy.empty((n_nodal_unknowns*n_nodes,n_data_cols))
+        for i_col in range(n_data_cols):
+            aux_col = numpy.array(self.snapshots_data_list[i_col])
+            snapshots_matrix[:,i_col] = aux_col.transpose()
+
+        # Calculate the randomized SVD of the snapshots matrix
+        u,_,_,_= RandomizedSingularValueDecomposition().Calculate(snapshots_matrix, self.svd_truncation_tolerance)
+
+        # Save the nodal basis
+        rom_basis_dict = {
+            "rom_settings": {},
+            "nodal_modes": {}
+        }
```

Comment:
This process looks good to me, the only question that comes to me now is. Do we want to include the "HROM" flag on this "rom_parameters.json"? 
In the case we want to, the user will need to modify this flag manually. 
What do you think @Rbravo555 ?

Reply:
I was wondering the same. What I'm not sure about is the default value, should we set it to `false` and let the user to manually change it if required?

Reply:
I would set it as `false` by default. And the name of the flag could be `train_hrom`. Because the HROM model is trained from the ROM model. And once the `ElementsAndWeights.json` file is created, the solver recognizes the existence of `hrom_elements` and automatically runs an HROM (the elements and weights are currently assigned in `MainKratos.py`). 

In this case, the sequence to run a HROM would be:
1.- Run FOM simulation with the training set using `calculate_rom_basis_output_process.py`.
2.- Manually, change `train_hrom` flag to `true` and run ROM simulation with the same training set.
3.- Once the `ElementsAndWeights.json` file is created, change the `train_hrom` flag to false and add to the `MainKratos.py` the necessary lines to assign the weights to the elements.
4.- Run HROM simulation.

Observe that we could also add a `assign_elements_and_weights_process.py` to avoid writing it on `MainKratos.py`.
Nevertheless, I've thought of another problem with the processes. When training ROM (run FOM), training HROM (run ROM) and/or running HROM, we would like to use the same `ProjectParameters.json` without building a new basis (without using `calculate_rom_basis_output_process`). Should we ask the user to erase this process? How could we tackle this problem? 

I wrote a lot, please let me know if I made myself clear.

Reply:
> I would set it as `false` by default. And the name of the flag could be `train_hrom`. Because the HROM model is trained from the ROM model. 

This makes perfect sense to me. I'd add it to the `rom_settings` set up. 

>
>And once the `ElementsAndWeights.json` file is created, the solver recognizes the existence of `hrom_elements` and automatically runs an HROM (the elements and weights are currently assigned in `MainKratos.py`).
> 

Maybe I didn't get your point at all, but if your suggestion is to filter if the simulation wheter uses hyper-reduction or not by the presence of the `ElementsAndWeights.json` I'm not full convinced. My reasoning is that you may want to run the same simulation with and without HROM, specially while developing. If we follow your suggestion this would require to constantly move the `ElementsAndWeights.json` file inside and outside the folder.

Instead, my suggestion would be to add another flag `run_hrom` to the `rom_settings` , which is set as `false` by default like we do with the `train_hrom`. Note that this makes also possible to unify the current ROM and HROM analysis as we could filter at run time if the simulation is wheter ROM or HROM.

Besides this, I think we should get rid of the `ElementsAndWeights.json` to append the element/conditions weights to a new section of the `RomParameters.json`. On top of getting rid of an extra file that pollutes the simulation folder, this stores the weights together with the basis that was used to obtain them, something that in my opinion is crucial for future documentation.

> In this case, the sequence to run a HROM would be: 1.- Run FOM simulation with the training set using `calculate_rom_basis_output_process.py`. 2.- Manually, change `train_hrom` flag to `true` and run ROM simulation with the same training set. 3.- Once the `ElementsAndWeights.json` file is created, change the `train_hrom` flag to false and add to the `MainKratos.py` the necessary lines to assign the weights to the elements. 4.- Run HROM simulation.
> 

Then, the sequence would be
1. Run a FOM simulation to obtain a first `RomParameters.json` with the nodal ROM basis.
2. If HROM is to be performed, manually change the `train_hrom` flag and run the ROM simulation. This would append the weights to the already existent `RomParameters.json` file.
3. Manually change the `train_hrom` to `false` and the `run_hrom` to `true` and run the HROM simulation. 

> Observe that we could also add a `assign_elements_and_weights_process.py` to avoid writing it on `MainKratos.py`. Nevertheless, I've thought of another problem with the processes. When training ROM (run FOM), training HROM (run ROM) and/or running HROM, we would like to use the same `ProjectParameters.json` without building a new basis (without using `calculate_rom_basis_output_process`). Should we ask the user to erase this process? How could we tackle this problem?
> 

Some thoughts about this:
1. It is much easier and much less annoying to ask the user to manually edit the json than the code (the script), so the current solution is in my opinion more user-friendly.
2. Eventually, we can enhance robustness by looping the processes lists in the ROM analysis to check if there is an instance of this process. If this is the case, we can remove it and throw a warning.

Reply:
I liked this sequence, and I have implemented it on another PR #9504 

> Then, the sequence would be
> 
> 1. Run a FOM simulation to obtain a first `RomParameters.json` with the nodal ROM basis.
> 2. If HROM is to be performed, manually change the `train_hrom` flag and run the ROM simulation. This would append the weights to the already existent `RomParameters.json` file.
> 3. Manually change the `train_hrom` to `false` and the `run_hrom` to `true` and run the HROM simulation.